src/Socket.js

```diff
@@ -0,0 +1,104 @@
+import { Socket as PhoenixSocket } from 'phoenix';
+
+export const states = {
+  CONNECTING: 'CONNECTING',
+  CONNECTED: 'CONNECTED',
+  WAITING: 'WAITING',
+  READY: 'READY',
+  ERROR: 'ERROR'
+};
+
+export default class Socket {
+  constructor($exp, socketURL, errorhandler) {
+    this.$exp = $exp;
+    this.errorHandler = (er) => {
+      this.state = states.ERROR;
+      errorhandler(er);
+    };
+
+    this.participantId = generateId(40);
+    this.phoenix = new PhoenixSocket(socketURL, {
+      params: {
+        participant_id: this.participantId,
+        experiment_id: this.$exp.id
+      }
+    });
+    this.phoenix.onError(this.errorHandler);
+    this.phoenix.connect();
+    this.state = states.CONNECTING;
+  }
+
+  initialize() {
+    this.participantChannel = this.phoenix.channel(
+      `participant:${this.participantId}`,
+      {}
+    );
+
+    this.participantChannel.on('experiment_available', (payload) => {
+      this.variant = payload.variant;
+      this.chain = payload.chain;
+      this.realization = payload.realization;
+      this.state = states.CONNECTED;
+    });
+
+    this.participantChannel
+      .join()
+      // Note that `receive` functions are for receiving a *reply* from the server after you try to send it something, e.g. `join()` or `push()`.
+      // While `on` function is for passively listening for new messages initiated by the server.
+      // We still need to wait for the actual confirmation message of "experiment_available". So we do nothing here.
+      .receive('ok', () => {})
+      .receive('error', this.errorHandler)
+      .receive('timeout', this.errorHandler);
+  }
+
+  join() {
+    this.roomChannel = this.phoenix.channel(
+      `interactive_room:${this.$exp.id}:${this.chain}:${this.realization}`,
+      { participant_id: this.participantId }
+    );
+
+    this.roomChannel
+      .join()
+      .receive('ok', () => {
+        this.state = states.WAITING;
+      })
+      .receive('error', this.errorHandler)
+      .receive('timeout', this.errorHandler);
+
+    this.roomChannel.on('start_game', () => {
+      this.state = states.READY;
+    });
+
+    this.setUpSubscriptions();
```

Comment:
What happens when you call this function without an argument?

Reply:
oops, that would be a mistake.